compiler/optimizer/OMRSimplifierHandlers.cpp

```diff
@@ -5260,6 +5260,111 @@ static bool skipRemLowering(int64_t divisor, TR::Simplifier *s)
    }
 
 
+/* Reduce Ifcmp Greater-Than-Or-Equal to Equal in the following case. If
+* - the first child is an "and" node
+* - the second child of the "and" node is a constant N
+* - the second child of the ifcmple node is a constant N,
+* Then, we can reduce ifcmple to ifcmpeq.
+*
+* For example, convert
+* ifiucmple --> block_5
+*  iand
+*    iload  <...>
+*    iconst 16
+*  iconst 16
+*
+*  into
+*
+* ifiucmpeq --> block_5
+*  iand
+*    iload  <...>
+*    iconst 16
+*  iconst 16
+*
+* Since the constants are equal, the case of greater-than/less-than is impossible,
+* and we can reduce a GE/LE operation to an EQ operation. After the reduction is performed,
+* the test-under-mask optimization can be applied to the "ifcmpeq" node.
+*/
+class IfcmpReducer
+   {
+   public:
+   IfcmpReducer(TR::Node *node, TR::Simplifier *s) : _node(node), _s(s) {}
+
+   static TR::ILOpCodes getNewOpCode(TR::Node *node)
+      {
+      auto opCodeValue = node->getOpCodeValue();
+      if (opCodeValue == TR::ifbcmple || opCodeValue == TR::ifbcmpge)
+         {
+         return TR::ifbcmpeq;
+         }
+      else if (opCodeValue == TR::ifbucmple || opCodeValue == TR::ifbucmpge)
+         {
+         return TR::ifbucmpeq;
+         }
+      else if (opCodeValue == TR::ifscmple || opCodeValue == TR::ifscmpge)
+         {
+         return TR::ifscmpeq;
+         }
+      else if (opCodeValue == TR::ifsucmple || opCodeValue == TR::ifsucmpge)
+         {
+         return TR::ifsucmpeq;
+         }
+      else if (opCodeValue == TR::ificmple || opCodeValue == TR::ificmpge)
+         {
+         return TR::ificmpeq;
+         }
+      else if (opCodeValue == TR::ifiucmple || opCodeValue == TR::ifiucmpge)
+         {
+         return TR::ifiucmpeq;
+         }
+      else if (opCodeValue == TR::iflcmple || opCodeValue == TR::iflcmpge)
+         {
+         return TR::iflcmpeq;
+         }
+      else if (opCodeValue == TR::iflucmple || opCodeValue == TR::iflucmpge)
+         {
+         return TR::iflucmpeq;
+         }
+      else
+         {
+         TR_ASSERT(false, "Ifcmp reduction is not implemented for node [%p]\n", node);
+         }
+      }
+
+   template<typename T>
+   bool reducible() const
+      {
+      TR::Node * firstChild = _node->getFirstChild(), * secondChild = _node->getSecondChild();
+
+      if (firstChild->getOpCode().isAnd() &&
+         firstChild->getSecondChild()->getOpCode().isIntegralConst() &&
+         secondChild->getOpCode().isIntegralConst())
+         {
+         T constAnd = firstChild->getSecondChild()->getConst<T>();
+         T constIfcmp = secondChild->getConst<T>();
+
+         if (constAnd == constIfcmp)
+            return true;
```

Comment:
This may not be correct. For example `if ((x & 3) <= 3)` is not equivalent to `if ((x & 3) == 3)` (let `x == 1` and this becomes apparent). In fact for all `<=` cases you capture `if ((x & 3) <= 3)` can be reduced to `true`. Its the `>=` that are special.

Reply:
it would still be correct for LE if the constant is a power of 2, though seems like a weird scenario. I'm ok to not catch this case in this PR :) .